tests/apps/tests.py

```diff
@@ -528,3 +530,50 @@ def test_multiple_paths_explicit_path(self):
             with self.settings(INSTALLED_APPS=["nsapp.apps.NSAppConfig"]):
                 app_config = apps.get_app_config("nsapp")
                 self.assertEqual(app_config.path, self.app_path)
+
+
+class QueryPerformingAppTests(TestCase):
+    """
+    Tests related to accessing the database during apps.populate()
+    Refs #33143
+    """
+
+    databases = {"default", "other"}
+
+    expected_message = (
+        "Accessing the database during app initialization is discouraged."
+    )
+
+    @override_settings(
+        INSTALLED_APPS=["apps.query_performing_app.apps.QueryDefaultDatabaseAppConfig"]
+    )
+    def test_setup_warns_about_sql_queries_when_importing_app_modules(self):
+        """
+        Trigger a warning when accessing the default database during apps.populate()
+        """
+        with patch.multiple(apps, ready=False, loading=False, app_configs={}):
+            with self.assertWarnsMessage(RuntimeWarning, self.expected_message):
+                django.setup()
+
+    @override_settings(
+        INSTALLED_APPS=["apps.query_performing_app.apps.QueryOtherDatabaseAppConfig"]
+    )
+    def test_setup_warns_about_sql_queries_to_non_default_databases(self):
+        """
+        Trigger a warning when accessing a non-default database during apps.populate()
+        """
+        with patch.multiple(apps, ready=False, loading=False, app_configs={}):
+            with self.assertWarnsMessage(RuntimeWarning, self.expected_message):
+                django.setup()
+
+    @override_settings(
+        INSTALLED_APPS=["apps.query_performing_app.apps.QueryUsingCursorAppConfig"]
+    )
+    def test_setup_warns_about_sql_queries_using_raw_cursors(self):
+        """
+        Trigger a warning when accessing the database using cursors during
+        apps.populate()
+        """
+        with patch.multiple(apps, ready=False, loading=False, app_configs={}):
+            with self.assertWarnsMessage(RuntimeWarning, self.expected_message):
+                django.setup()
```

Comment:
Not really sure what you mean or how to do this... I mean, the warning is only issued if the app executes a query. But since its only a warning, the queries should be executed as usual.

I now just set a flag on the `AppConfig` instance whether the query completed successfully or not, but this feels pretty weird. Should I maybe create a model instance in `AppConfig.ready()` and check that its the same object afterwards?

Also, what about testing that the warning is _not_ triggered when we don't execute any queries? 😅

Reply:
> Also, what about testing that the warning is not triggered when we don't execute any queries? sweat_smile

All warnings are treated as errors in `runtests.py` so it's implicitly tested many times.

Reply:
> All warnings are treated as errors in `runtests.py` so it's implicitly tested many times.

> We should a add test with `ignore_warnings(category=RuntimeWarning)` to confirm that queries can still be executed.

Isn't `SimpleTestCase.assertWarnsMessage()` doing that already? I mean, it's recording the warnings instead of just ignoring them. But they don't turn into exceptions so the queries execute fine, don't they?

Reply:
> of just ignoring them. But they don't turn into exceptions so the queries execute fine, don't they?

Yes, but the previous version with caught exception etc. didn't actually check that the queries were executed properly. TBH, I'd prefer to check that the query returned expected results.

Reply:
Okay, I'm checking the query results now